keysync_responder.go

```diff
@@ -0,0 +1,127 @@
+package nitriding
+
+import (
+	cryptoRand "crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hf/nitrite"
+	"golang.org/x/crypto/nacl/box"
+)
+
+var (
+	errFailedNonce     = "failed to create nonce"
+	errNoBase64        = "failed to Base64-decode attestation document"
+	errFailedVerify    = "failed to verify attestation document"
+	errFailedRespBody  = "failed to read response body"
+	errFailedPCR       = "failed to get PCR values"
+	errFailedFindNonce = "could not find provided nonce"
+	errInvalidBoxKeys  = "invalid box key material"
+	errPCRNotIdentical = "remote enclave's PCR values not identical"
+)
+
+type timeFunc func() time.Time
+
+// getNonceHandler returns a HandlerFunc that creates a new nonce and returns
+// it to the client.
+func getNonceHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		nonce, err := newNonce()
+		if err != nil {
+			http.Error(w, errFailedNonce, http.StatusInternalServerError)
+			return
+		}
+
+		e.nonceCache.Add(nonce.B64())
+		fmt.Fprintf(w, "%s\n", nonce.B64())
+	}
+}
+
+func getKeysHandler(e *Enclave, curTime timeFunc) http.HandlerFunc {
```

Comment:
would be nice if this had a doc comment too.

Reply:
Fixed in db91b5a.